src/pint/models/model_builder.py

```diff
@@ -817,3 +915,149 @@ def get_model_and_toas(
         mm.add_tzr_toa(tt)
 
     return mm, tt
+
+
+def guess_binary_model(parfile_dict):
+    """Based on the PINT parameter dictionary, guess the binary model
+
+    Parameters
+    ----------
+    parfile_dict
+        The parameter dictionary as read-in by parse_parfile
+
+    Returns
+    -------
+    list:
+        A priority-ordered list of possible binary models. The first one is the
+        best-guess
+
+    """
+
+    def add_sini(parameters):
+        """If 'KIN' is a model parameter, Tempo2 doesn't really use SINI"""
+        if "KIN" in parameters:
+            return list(parameters) + ["SINI"]
+        else:
+            return list(parameters)
+
+    all_components = AllComponents()
+    binary_models = all_components.category_component_map["pulsar_system"]
+
+    # Find all binary parameters
+    # TODO: Check whetehr binary_model[:6]=="Binary" ??
+    binary_parameters_map = {
+        binary_model[6:]: add_sini(
+            all_components.search_binary_components(binary_model).aliases_map.keys()
+        )
+        for binary_model in binary_models
+    }
+    binary_parameters_map.update({"Isolated": []})
+    all_binary_parameters = {
+        parname for parnames in binary_parameters_map.values() for parname in parnames
+    }
+
+    # Find all parfile parameters
+    all_parfile_parameters = set(parfile_dict.keys())
+
+    # All binary parameters in the parfile
+    parfile_binary_parameters = all_parfile_parameters & all_binary_parameters
+
+    # Find which binary models include those
+    allowed_binary_models = {
+        binary_model
+        for (binary_model, bmc) in binary_parameters_map.items()
+        if len(parfile_binary_parameters - set(bmc)) == 0
+    }
+
+    # The ordering/priority of the binary models
+    binary_model_priority = [
+        "Isolated",
+        "BT",
+        "ELL1",
+        "ELL1H",
+        "ELL1k",
+        "DDK",
+        "DDGR",
+        "DDS",
+        "DD",
+    ]
+
```

Comment:
Maybe add an assertion here about the `binary_model_priority` list including *all* implemented binary models? This way it will throw an error if somebody implements a new timing model and doesn't update this list.

Reply:
Is an assertion at this place the right place for it? I could also move the `binary_model_priority` somewhere more central (any ideas where?), and then do the assertion in the `test_t2binary2pint` file for instance.

Reply:
Yes... The assertion may be better placed in a test script. 

Maybe `binary_model_priority` can be a private global variable (`_binary_model_priority`) within `model_builder.py`?

Reply:
Done